lib/crypto/hkdf.lua

```diff
@@ -0,0 +1,115 @@
+end
+
+--- Creates a new HKDF instance for a given hash algorithm.
+-- @function HKDF.new
+-- @tparam string alg The base hash algorithm name (e.g., "sha256", "sha512").
+-- The "hmac(" prefix will be added automatically.
+-- @treturn HKDF An HKDF instance table with methods for key derivation.
+-- @usage local hkdf_sha256 = require("crypto.hkdf").new("sha256")
+function HKDF.new(alg)
+	local hmac = setmetatable({}, HKDF)
+	local ok, tfm = pcall(shash.new, "hmac(" .. alg .. ")")
+	assert(ok, "Invalid hash algorithm: " .. alg)
+	hmac.tfm = tfm
+	hmac.salt = rep("\0", #hmac)
+
+	return hmac
+end
+
+--- Performs an HMAC calculation using the instance's algorithm.
+-- @tparam string key The HMAC key.
+-- @tparam string data The data to hash.
+-- @treturn string The HMAC digest.
+function HKDF:hmac(key, data)
+	self.tfm:setkey(key)
+	return self.tfm:digest(data)
+end
+
+--- Performs the HKDF Extract step.
+-- @function HKDF:extract
+-- @tparam[opt] string salt Optional salt value. If nil or not provided, a salt of `hash_len` zeros is used.
+-- @tparam string ikm Input Keying Material.
+-- @treturn string The Pseudorandom Key (PRK).
+function HKDF:extract(salt, ikm)
+	return self:hmac((salt or self.default_salt), ikm)
+end
+
+--- Performs the HKDF Expand step.
+-- @function HKDF:expand
+-- @tparam string prk Pseudorandom Key.
+-- @tparam[opt] string info Optional context and application-specific information. Defaults to an empty string if nil.
+-- @tparam number length The desired length in bytes for the Output Keying Material (OKM).
+-- @treturn string The Output Keying Material of the specified `length`.
+function HKDF:expand(prk, info, length)
+	info = info or ""
+	local hash_len = #self
+	local n = length / hash_len
+	n = (n * hash_len == length) and n or n + 1
+
+	local okm, t = "", ""
+	for i = 1, n do
+		t = self:hmac(prk, t .. info .. char(i))
+		okm = okm .. t
+	end
+	return sub(okm, 1, length)
+end
+
+--- Performs the full HKDF (Extract and Expand) operation.
+-- @function HKDF:hkdf
+-- @tparam[opt] string salt Optional salt value.
+-- @tparam string ikm Input Keying Material.
+-- @tparam[opt] string info Optional context and application-specific information.
+-- @tparam number length The desired length in bytes for the Output Keying Material.
+-- @treturn string The Output Keying Material.
+function HKDF:hkdf(salt, ikm, info, length)
+	return self:expand(self:extract(salt, ikm), info, length)
+end
+
+--- Derives a key using HKDF-Expand-Label as defined in RFC 8446 for TLS 1.3.
+-- @function HKDF:tls13_expand_label
+-- @tparam string prk Pseudorandom Key.
+-- @tparam string label The label for the derived secret.
+-- @tparam string context A string containing the hash of the transcript of the handshake messages.
+-- @tparam number length The desired length in bytes for the derived secret.
+-- @treturn string The derived secret of the specified `length`.
+function HKDF:tls13_expand_label(prk, label, context, length)
```

Comment:
it looks to me that this function is more like a client of HKDF API than a feature from it.. I'd move it to a tls3 module and scope it out from this PR..